clients/filesystem-hadoop/src/main/java/com/datastrato/gravitino/filesystem/hadoop/GravitinoVirtualFileSystem.java

```diff
@@ -0,0 +1,406 @@
+            + identifier.namespace().level(2)
+            + "/"
+            + identifier.name();
+    if (meta.getFileset().storageLocation().endsWith("/")) {
+      filesetPrefix += "/";
+    }
+    return filesetPrefix;
+  }
+
+  @SuppressWarnings("deprecation")
+  private Path resolvePathByIdentifier(NameIdentifier identifier, FilesetMeta meta, Path path) {
+    String originPath = path.toString();
+    if (!originPath.startsWith(GravitinoVirtualFileSystemConfiguration.GVFS_FILESET_PREFIX)) {
+      throw new InvalidPathException(
+          String.format(
+              "Path %s doesn't start with the scheme \"%s\".",
+              path, GravitinoVirtualFileSystemConfiguration.GVFS_FILESET_PREFIX));
+    }
+    try {
+      Path storageLocation = new Path(meta.getFileset().storageLocation());
+      boolean isMountSingleFile = meta.getFileSystem().isFile(storageLocation);
+      if (isMountSingleFile) {
+        Preconditions.checkArgument(
+            originPath.equals(concatFilesetPrefix(identifier, meta)),
+            "Cannot resolve path: %s to actual storage path, because the fileset only mounts a single file.",
+            path);
+        return storageLocation;
+      } else {
+        return new Path(
+            originPath.replaceFirst(
+                concatFilesetPrefix(identifier, meta), storageLocation.toString()));
+      }
+    } catch (Exception e) {
+      throw new RuntimeException(
+          String.format("Cannot resolve path: %s to actual storage path, exception:", path), e);
+    }
+  }
+
+  private FileStatus resolveFileStatusPathPrefix(
+      FileStatus fileStatus, String fromPrefix, String toPrefix) {
+    String filePath = fileStatus.getPath().toString();
+    if (!filePath.startsWith(fromPrefix)) {
+      throw new InvalidPathException(
+          String.format("Path %s doesn't start with prefix \"%s\".", filePath, fromPrefix));
+    }
+    String proxyPath = filePath.replaceFirst(fromPrefix, toPrefix);
+    Path path = new Path(proxyPath);
+    fileStatus.setPath(path);
+    return fileStatus;
+  }
+
+  private NameIdentifier extractIdentifier(URI proxyUri) {
+    Preconditions.checkArgument(
+        proxyUri.toString().startsWith(GravitinoVirtualFileSystemConfiguration.GVFS_FILESET_PREFIX),
+        "Path %s doesn't start with scheme \"%s\".");
+
+    if (StringUtils.isBlank(proxyUri.toString())) {
+      throw new InvalidPathException("Uri which need be extracted cannot be null or empty.");
+    }
+
+    // remove first '/' symbol with empty string
+    String[] reservedDirs =
+        Arrays.stream(proxyUri.getPath().replaceFirst("/", "").split("/")).toArray(String[]::new);
+    Preconditions.checkArgument(
+        reservedDirs.length >= 3, "URI %s doesn't contains valid identifier", proxyUri);
+
+    return NameIdentifier.ofFileset(
+        metalakeName, reservedDirs[0], reservedDirs[1], reservedDirs[2]);
+  }
+
+  private FilesetMeta getCachedFileset(NameIdentifier identifier) throws IOException {
```

Comment:
I think we can build a cache method to create a new fileset meta if not existed. for example: `cache.get(id, () -> createNewCache)`, so that caffeine will guarantee the atomicity.